src/ast/visitor.rs

```diff
@@ -0,0 +1,148 @@
+use crate::{
+    ast::{BinOp, Block, Literal, Located, Name, Node, UnOp},
+    ty::{Binding, Ty},
+};
+
+pub trait NodeVisitor<'a> {
+    fn super_block(&mut self, block: &Block<'a>) {
+        for node in block {
+            self.visit_node(&node);
+        }
+    }
+
+    fn super_node(&mut self, node: &Located<Node<'a>>) {
+        match &node.content {
+            Node::BinaryOp(op, node1, node2) => {
+                self.visit_binary_op(&op, node1.as_ref(), node2.as_ref())
+            }
+            Node::UnaryOp(op, node) => self.visit_unary_op(&op, node.as_ref()),
+            Node::LetBind(name, opt_ty, node) => self.visit_let_bind(&name, &opt_ty, node.as_ref()),
+            Node::Cond(if_blk, do_blk, el_blk) => self.visit_cond(&if_blk, &do_blk, &el_blk),
+            Node::FnDef(opt_name, args, body, opt_ty) => {
+                self.visit_fn_def(&opt_name, &args, &body, &opt_ty)
+            }
+            Node::Call(func, args) => self.visit_call(&func, &args),
+            Node::Literal(literal) => self.visit_literal(&literal),
+            Node::Name(name) => self.visit_name(&name),
+        }
+    }
+
+    fn super_binary_op(
+        &mut self,
+        _op: &BinOp,
```

Comment:
BinOp can be passed per value (size_of BinOP < sizeof reference)

Reply:
Thanks! I've done it now :)